src/sentry/debug_files/artifact_bundle_indexing.py

```diff
@@ -227,70 +171,146 @@ def get_flat_file_meta(self) -> Optional[FlatFileMeta]:
 
         return meta
 
-    def _locking_key(self):
-        return f"bundle_index:write:{self._hashed()}"
 
-    def get_lock(self) -> Lock:
-        locking_key = self._locking_key()
+@sentry_sdk.tracing.trace
+def mark_bundle_for_flat_file_indexing(
+    artifact_bundle: ArtifactBundle,
+    project_ids: List[int],
+    release: Optional[str],
+    dist: Optional[str],
+) -> List[FlatFileIdentifier]:
+    identifiers = []
+
+    for project_id in project_ids:
+        if release:
+            identifiers.append(
+                FlatFileIdentifier(project_id, release=release, dist=dist or NULL_STRING)
+            )
+
+        identifiers.append(FlatFileIdentifier.for_debug_id(project_id))
+
+    # Create / Update the indexing state in the database
+    for identifier in identifiers:
+        with atomic_transaction(
+            using=(
+                router.db_for_write(ArtifactBundleFlatFileIndex),
+                router.db_for_write(FlatFileIndexState),
+            )
+        ):
+            flat_file_index, _created = ArtifactBundleFlatFileIndex.objects.get_or_create(
+                project_id=identifier.project_id,
+                release_name=identifier.release,
+                dist_name=identifier.dist,
+            )
+            FlatFileIndexState.objects.update_or_create(
```

Comment:
The problem I find with leveraging uniqueness constraints failing, is that insertions will fail on conflict and will eventually be fixed in the future. It would be so nice to be able to serialize access so that we can process without race conditions.